### PR TITLE
Per program configurations

### PR DESCRIPTION
**Pull request by lalithsuresh**
Is this a user-visible change (yes/no): yes

### CODE REVIEW

**Comment by gz** [comment]
I think this should be configurable for a pipeline, rather than per program.

**Comment by lalithsuresh** [comment]
@gz Leonid and I discussed this. Unfortunately we compile programs independently of pipelines today.

File: crates/pipeline_manager/src/compiler.rs [hunk 1]
```diff
@@ -321,7 +328,8 @@ impl Compiler {
             })?;
 
         // `cargo build`.
-        let mut cargo_process = Compiler::run_cargo_build(config, program_id).await?;
+        let mut cargo_process =
+            Compiler::run_cargo_build(config, program_id, &CompilationProfile::Optimized).await?;
```
[the remark this answers is not in the document]
Comment (lalithsuresh) [reply crates/pipeline_manager/src/compiler.rs:332]: Yes. Wasn't sure whether the additional image size bloat was worth the additional precompile. For what it's worth, I think cold unoptimized builds were taking about as long as a warm optimized build.

**Review by ryzhyk — APPROVED** [review]

File: crates/pipeline_manager/src/db/program.rs [hunk 2]
```diff
@@ -263,18 +282,29 @@ pub(crate) async fn update_program(
...
+                                           ELSE NULL END),
+                            compilation_profile = $12
                     WHERE tenant_id = $1 AND id = $2
                     RETURNING version
                 "#,
             )
             .await?;
     let has_code_changed = program_code.as_ref().is_some_and(|c| *c != code);
-    let new_version = if has_code_changed {
+    let has_config_changed = config.as_ref().is_some_and(|c| c.profile != profile);
```
Comment (snkas) [inline crates/pipeline_manager/src/db/program.rs:293]: `has_profile_changed`

File: crates/pipeline_manager/src/db/program.rs [hunk 3]
```diff
@@ -263,18 +282,29 @@ pub(crate) async fn update_program(
...
+
+    // Changing the program configuration currently counts as a new version, which in turn
```
Comment (snkas) [inline crates/pipeline_manager/src/db/program.rs:295]: ... the program compilation profile currently ...

**Review by snkas — APPROVED** [review]
Looks good, I think it will be a very useful feature.

**Pull request merged**